### PR TITLE
Add VII L2 netCDF-reader.

### PR DESCRIPTION
**Pull request by sjoro**
Add EPS-SG VII L2 netCDF reader. Replaces PR #1074 and is based on a most recent master-branch.

 - [x] Tests added
 - [x] Tests passed
 - [x] Passes ``flake8 satpy``
 - [x] Fully documented

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,2 @@
+# epssg-satpy
```
Comment (mraspaud) [inline README.md]: Should this be here ?

File: satpy/readers/netcdf_utils.py [hunk 2]
```diff
@@ -143,13 +146,15 @@ def collect_metadata(self, name, obj):
...
+            full_group_name = base_name + group_name
+            self.file_content[full_group_name] = group_obj
+            self._collect_attrs(full_group_name, group_obj)
+            self.collect_metadata(full_group_name, group_obj)
         for var_name, var_obj in obj.variables.items():
             var_name = base_name + var_name
             self.file_content[var_name] = var_obj
             self.file_content[var_name + "/dtype"] = var_obj.dtype
             self.file_content[var_name + "/shape"] = var_obj.shape
-            self.file_content[var_name + "/dimensions"] = var_obj.dimensions
```
Comment (mraspaud) [inline satpy/readers/netcdf_utils.py:152]: This was added a couple of weeks ago, why remove it ?

File: satpy/readers/netcdf_utils.py [hunk 3]
```diff
@@ -143,13 +146,15 @@ def collect_metadata(self, name, obj):
         # Look through each subgroup
         base_name = name + "/" if name else ""
         for group_name, group_obj in obj.groups.items():
-            self.collect_metadata(base_name + group_name, group_obj)
+            full_group_name = base_name + group_name
+            self.file_content[full_group_name] = group_obj
+            self._collect_attrs(full_group_name, group_obj)
+            self.collect_metadata(full_group_name, group_obj)
```
Comment (mraspaud) [inline satpy/readers/netcdf_utils.py:152]: @gerritholl you worked with this last, would you mind checking this change?

File: satpy/readers/netcdf_utils.py [hunk 4]
```diff
@@ -196,6 +202,11 @@ def __getitem__(self, key):
                 val = self._get_var_from_filehandle(group, key)
             else:
                 val = self._get_var_from_xr(group, key)
+        elif isinstance(val, netCDF4.Group):
+            # Full groups are conveniently read with xr even if file_handle is available
+            with xr.open_dataset(self.filename, group=key,
+                                 **self._xarray_kwargs) as nc:
+                val = nc
```
Comment (mraspaud) [inline satpy/readers/netcdf_utils.py:209]: Will the context manager not close the nc file? Will we not loose access to the data if we use dask in the background?

File: satpy/readers/netcdf_utils.py [hunk 5]
```diff
@@ -217,10 +228,7 @@ def _get_var_from_xr(self, group, key):
     def _get_var_from_filehandle(self, group, key):
         # Not getting coordinates as this is more work, therefore more
         # overhead, and those are not used downstream.
-        if group is None:
-            g = self.file_handle
-        else:
-            g = self.file_handle[group]
+        g = self.file_handle[group]
```
Comment (mraspaud) [inline satpy/readers/netcdf_utils.py:231]: From what I see `group` can still be None: https://github.com/pytroll/satpy/pull/1219/files#diff-cfed382dd19ba8d52b7734cecf817407R197-R200
Is that an accepted key ?

File: satpy/readers/vii_utils.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+"""Utilities for the management of VII products."""
+
+
+# PLANCK COEFFICIENTS FOR CALIBRATION AS DEFINED BY EUMETSAT
+C1 = 1.191062e+8   # [W/m2·sr-1·µm4]
+C2 = 1.4387863e+4  # [K·µm]
+
+# CONSTANTS DEFINING THE TIE POINTS
+TIE_POINTS_FACTOR = 8    # Sub-sampling factor of tie points wrt pixel points
+SCAN_ALT_TIE_POINTS = 4  # Number of tie points along the satellite track for each scan
```
Comment (mraspaud) [inline satpy/readers/vii_utils.py:28]: Isn't this defined in the data file???

File: satpy/tests/reader_tests/test_netcdf_utils.py [hunk 7]
```diff
@@ -54,7 +54,6 @@ def get_test_content(self, filename, filename_info, filetype_info):
             - '/attr/global_attr'
             - 'dataset/attr/global_attr'
             - 'dataset/shape'
-            - 'dataset/dimensions'
```
Comment (mraspaud) [inline satpy/tests/reader_tests/test_netcdf_utils.py:57]: Looks like you are reverting changes done recently here.

File: satpy/readers/vii_utils.py [hunk 8]
```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# Copyright (c) 2019 Satpy developers
```
Comment (mraspaud) [inline satpy/readers/vii_utils.py:4]: It's 2020 now :)

File: satpy/readers/vii_l2_nc.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+#
+# satpy is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with satpy.  If not, see <http://www.gnu.org/licenses/>.
+
+"""EUMETSAT EPS-SG Visible/Infrared Imager (VII) Level 2 products reader."""
```
Comment (mraspaud) [inline satpy/readers/vii_l2_nc.py:19]: A link to the documentation for the format should be added here.

File: satpy/readers/vii_base_nc.py [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+
+        Args:
+            variable: xarray DataArray containing the dataset to wrap.
+
+        Returns:
+            DataArray: array containing the wrapped values and all the original metadata
+
+        """
+        variable %= 360.
+        return variable
```
Comment (mraspaud) [inline satpy/readers/vii_base_nc.py:113]: This function is used once and is just a wrapper around `%= 360.` Why is this needed ?

File: satpy/readers/vii_base_nc.py [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+from satpy.readers.vii_utils import SCAN_ALT_TIE_POINTS, TIE_POINTS_FACTOR
+from geotiepoints.viiinterpolator import tie_points_interpolation, tie_points_geo_interpolation
+
+logger = logging.getLogger(__name__)
+
+
+class ViiNCBaseFileHandler(NetCDF4FileHandler):
+    """Base reader class for VII products in netCDF format."""
+
+    def __init__(self, filename, filename_info, filetype_info, orthorect=False):
```
Comment (mraspaud) [inline satpy/readers/vii_base_nc.py:35]: The orthorect parameter isn't documented.

File: satpy/readers/netcdf_utils.py (hunk 2 above)
Comment (gerritholl) [reply satpy/readers/netcdf_utils.py:152]: @sjoro Are you sure you used the latest master?  Both here and at the change around (new) line 230-231 the PR appears to revert changes that were merged into master recently, see https://github.com/pytroll/satpy/pull/1195 and https://github.com/pytroll/satpy/pull/1199.

File: README.md (hunk 1 above)
Comment (sjoro) [reply README.md]: removed.

File: satpy/readers/netcdf_utils.py (hunk 2 above)
Comment (sjoro) [reply satpy/readers/netcdf_utils.py:152]: the `netcdf_utils`-file in this PR is not from the latest master.

File: satpy/readers/netcdf_utils.py (hunk 4 above)
Comment (sjoro) [reply satpy/readers/netcdf_utils.py:209]: From my tests, the problem does not happen. Data can be correctly read.

File: satpy/readers/vii_l2_nc.py (hunk 9 above)
Comment (sjoro) [reply satpy/readers/vii_l2_nc.py:19]: not publicly available yet. maybe put a TODO here?

File: satpy/readers/vii_utils.py (hunk 6 above)
Comment (sjoro) [reply satpy/readers/vii_utils.py:28]: not directly. indirectly it is possible by utilizing some other variables in the file and then doing some computations/inspections. which one to implement? as is or using the variables from the file?

File: satpy/tests/reader_tests/test_netcdf_utils.py (hunk 7 above)
Comment (sjoro) [reply satpy/tests/reader_tests/test_netcdf_utils.py:57]: same as earlier, `netcdf_utils`in thi PR is not from the latest Satpy-master.

File: satpy/readers/vii_l2_nc.py (hunk 9 above)
Comment (sjoro) [reply satpy/readers/vii_l2_nc.py:19]: issue #1231 raised.

File: satpy/readers/netcdf_utils.py (hunk 3 above)
Comment (gerritholl) [reply satpy/readers/netcdf_utils.py:152]: Looks good to me.  I originally introduced this for performance reasons, I observe no performance difference in the script I used to measure this back then.

Comment (mraspaud) [reply satpy/readers/netcdf_utils.py:152]: Thanks for checking it out

**Comment by mraspaud** [comment]
Trigger rebuild
